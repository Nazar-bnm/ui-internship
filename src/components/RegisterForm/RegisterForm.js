import React, { Component } from 'react';
import ls from 'local-storage';
import cx from 'classnames';

import { Link } from 'react-router-dom';
import { CountryDropdown, RegionDropdown } from 'react-country-region-selector';
import { Button } from '@/shared';
import { VALIDATION_FAILED, REGISTER_FAILED, REGISTER_SUCCESS } from '../../constants/notificationData';

import './RegisterForm.scss';
import HttpService from '@/service/HttpService/httpService';

const CN = 'register-form';

class RegisterForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      address: {
        company: '',
        addressOne: '',
        country: '',
        city: '',
        postalCode: '',
        phone: ''
      }
    };

    this.formRegister = React.createRef();
    this.validate = this.validate.bind(this);
    this.sendData = this.sendData.bind(this);
  }

  componentDidMount() {
    this.setState({
      email: ls.get('emailToRegister') || ''
    });
  }

  setAddressValue(value, fieldName) {
    const { address } = this.state;
    const copiedAddress = { ...address };

    if (fieldName === 'postalCode') {
      copiedAddress[fieldName] = +value;
      this.setState({ address: { ...copiedAddress } });
    } else
    if (fieldName === 'phone') {
      copiedAddress[fieldName] = +value;
      this.setState({ address: { ...copiedAddress } });
    } else {
      copiedAddress[fieldName] = value;
      this.setState({ address: { ...copiedAddress } });
    }
  }

  setFirstName(val) {
    this.setState({ firstName: val });
  }

  setLastName(val) {
    this.setState({ lastName: val });
  }

  setEmail(val) {
    this.setState({ email: val });
  }

  setPassword(val) {
    this.setState({ password: val });
  }

  validate() {
    const { showMessage } = this.props;

    let isInputValid = false;

    isInputValid = this.formRegister.current.reportValidity();

    if (!isInputValid) {
      showMessage(VALIDATION_FAILED);
    } else {
      this.sendData();
    }

    return isInputValid;
  }


  async sendData() {
    const { showMessage } = this.props;
    const userAPI = new HttpService();
    const signupURL = `${process.env.SERVER_URL}/signup`;
    const {
      firstName,
      lastName,
      email,
      password,
      address
    } = this.state;
    const userData = {
      firstName,
      lastName,
      email,
      password,
      address
    };

    try {
      const response = await userAPI.post(signupURL, userData);
      const errorMessage = response.data.message;

      if (errorMessage) {
        showMessage(REGISTER_FAILED);
      } else {
        showMessage(REGISTER_SUCCESS);
      }
    } catch (error) {
      showMessage(REGISTER_FAILED);
    }
  }


  render() {
    const { email, address: { country, region } } = this.state;

    return (
      <form
        className={cx(CN)}
        onSubmit={(e) => e.preventDefault()}
        ref={this.formRegister}
      >
        <span className={`${CN}__label`}>create an account</span>
        <div className={`${CN}__wrapper`}>
          <div className={`${CN}__personal-information ${CN}__wrapper__personal-information`}>

            <h3>personal information</h3>
            <div className={`${CN}__field`}>
              <label htmlFor="firstName" className={`${CN}__field__label`}>first name*</label>
              <input
                name="firstName"
                className={`${CN}__field__input`}
                type="text"
                required
                onChange={(evt) => this.setFirstName(evt.target.value)}
              />
            </div>

            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>last name*</label>
              <input
                name="lastName"
                className={`${CN}__field__input`}
                type="text"
                required
                onChange={(evt) => this.setLastName(evt.target.value)}
              />
            </div>

            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>e-mail*</label>
              <input
                name="email"
                className={`${CN}__field__input`}
                type="email"
                id="email"
                defaultValue={email}
                required
                onChange={(evt) => this.setEmail(evt.target.value)}
              />
            </div>

            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>password*</label>
              <input
                name="password"
                className={`${CN}__field__input`}
                type="password"
                required
                pattern="^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$"
                onChange={(evt) => this.setPassword(evt.target.value)}
              />
            </div>
          </div>

          <div className={`${CN}__address-information ${CN}__wrapper__address-information`}>

            <h3>address information</h3>
            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>company</label>
              <input className={`${CN}__field__input`} type="text" onChange={(evt) => this.setAddressValue(evt.target.value, 'company')} />
            </div>

            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>address</label>
              <input className={`${CN}__field__input`} type="text" onChange={(evt) => this.setAddressValue(evt.target.value, 'addressOne')} />
            </div>

            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>country</label>
              <CountryDropdown
                className={`${CN}__field__country`}
                value={country}
                onChange={(evt) => this.setAddressValue(evt, 'country')}
              />
            </div>

            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>region</label>
              <RegionDropdown
                className={`${CN}__field__country`}
                country={country}
                value={region}
                onChange={(evt) => this.setAddressValue(evt, 'region')}
              />
            </div>

            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>city</label>
              <input className={`${CN}__field__input`} type="text" onChange={(evt) => this.setAddressValue(evt.target.value, 'city')} />
            </div>

            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>zip/postal code</label>
              <input className={`${CN}__field__input`} type="number" onChange={(evt) => this.setAddressValue(evt.target.value, 'postalCode')} />
            </div>

            <div className={`${CN}__field`}>
              <label className={`${CN}__field__label`}>phone</label>
              <input className={`${CN}__field__input`} type="number" onChange={(evt) => this.setAddressValue(evt.target.value, 'phone')} />
            </div>
          </div>
        </div>

        <div className={`${CN}__buttons-wrapper`}>
          <Button
            type="submit"
            className="black-button"
            onClick={this.validate}
          >
            create an account
          </Button>
          <Link className="link-back" to="/login">
            Back to login
          </Link>
        </div>
      </form>
    );
  }
}

export default RegisterForm;
