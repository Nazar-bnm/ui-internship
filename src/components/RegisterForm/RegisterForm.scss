@import '../../scss/global_styles';

.register-form {
  margin-top: 20px;
  border: 1px solid $silver-chalice;
  width: 60%;
  margin: auto;

  &__label {
    display: block;
    text-transform: uppercase;
    font-family: font($type: secondary);
    font-size: 1.5rem;
    text-align: center;
    border-bottom: 1px solid $silver-chalice;
    padding: 10px;
  }

  &__wrapper {
    @include flex(row, space-between, null);

    &__personal-information,
    &__address-information {
      margin: 40px 0;
      text-transform: uppercase;
      font-family: font($type: primary-light);
      font-size: 0.8rem;
    }
    
    width: 90%;
    margin: auto;
    padding-left: 10px;

    .errorMessage {
      color: red;
      font-size: 0.75em;
      display: relative;
    }
  }

  &__field {
    display: flex;
    flex-direction: column;
    margin-top: 10px;
    font-family: font($type: primary-light);
    font-size: 0.6rem;

    &__input {
      margin-top: 3px;
      height: 2rem;
      width: 20vw;
      padding-left: 10px;
      border: 1px solid $silver-chalice;
    }

    &__country {
      width: 20vw;
      height: 30px;
      font-size: inherit;
      background-color: white;
      border-radius: 0%;
      
    }
  }

  &__buttons-wrapper {
    @include flex(column, center, center);

    .black-button {
      width: 20vh;
      height: 40px;
      padding: 10px;
      margin-right: 0;
    }
    
    .link-back {
      margin: 10px 0;
      color: black
    }
  }
}