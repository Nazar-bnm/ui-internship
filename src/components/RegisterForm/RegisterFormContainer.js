import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import RegisterForm from './RegisterForm';
import showMessage from '../../actions/actionsNotifications';

const mapDispatchToProps = (dispatch) => ({
  showMessage: bindActionCreators(showMessage, dispatch)
});

export default connect(null, mapDispatchToProps)(RegisterForm);
