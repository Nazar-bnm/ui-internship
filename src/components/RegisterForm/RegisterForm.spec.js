import React from 'react';
import { shallow } from 'enzyme';

import RegisterForm from './RegisterForm';

describe('<RegisterForm />', () => {
  test('should match snapshot', () => {
    const wrapper = shallow(<RegisterForm />);
    expect(wrapper).toMatchSnapshot();
  });
});
