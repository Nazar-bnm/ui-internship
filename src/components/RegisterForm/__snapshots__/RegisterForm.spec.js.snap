// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterForm /> should match snapshot 1`] = `
<form
  className="register-form"
  onSubmit={[Function]}
>
  <span
    className="register-form__label"
  >
    create an account
  </span>
  <div
    className="register-form__wrapper"
  >
    <div
      className="register-form__personal-information register-form__wrapper__personal-information"
    >
      <h3>
        personal information
      </h3>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
          htmlFor="firstName"
        >
          first name*
        </label>
        <input
          className="register-form__field__input"
          name="firstName"
          onChange={[Function]}
          required={true}
          type="text"
        />
      </div>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          last name*
        </label>
        <input
          className="register-form__field__input"
          name="lastName"
          onChange={[Function]}
          required={true}
          type="text"
        />
      </div>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          e-mail*
        </label>
        <input
          className="register-form__field__input"
          defaultValue=""
          id="email"
          name="email"
          onChange={[Function]}
          required={true}
          type="email"
        />
      </div>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          password*
        </label>
        <input
          className="register-form__field__input"
          name="password"
          onChange={[Function]}
          pattern="^(?=.*[A-Za-z])(?=.*\\\\d)[A-Za-z\\\\d]{8,}$"
          required={true}
          type="password"
        />
      </div>
    </div>
    <div
      className="register-form__address-information register-form__wrapper__address-information"
    >
      <h3>
        address information
      </h3>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          company
        </label>
        <input
          className="register-form__field__input"
          onChange={[Function]}
          type="text"
        />
      </div>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          address
        </label>
        <input
          className="register-form__field__input"
          onChange={[Function]}
          type="text"
        />
      </div>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          country
        </label>
        <CountryDropdown
          blacklist={Array []}
          className="register-form__field__country"
          classes=""
          defaultOptionLabel="Select Country"
          disabled={false}
          id=""
          labelType="full"
          name="rcrs-country"
          onBlur={[Function]}
          onChange={[Function]}
          priorityOptions={Array []}
          showDefaultOption={true}
          value=""
          valueType="full"
          whitelist={Array []}
        />
      </div>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          region
        </label>
        <RegionDropdown
          blankOptionLabel="-"
          className="register-form__field__country"
          classes=""
          country=""
          countryValueType="full"
          customOptions={Array []}
          defaultOptionLabel="Select Region"
          disableWhenEmpty={false}
          disabled={false}
          id=""
          labelType="full"
          name="rcrs-region"
          onBlur={[Function]}
          onChange={[Function]}
          showDefaultOption={true}
          value=""
          valueType="full"
        />
      </div>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          city
        </label>
        <input
          className="register-form__field__input"
          onChange={[Function]}
          type="text"
        />
      </div>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          zip/postal code
        </label>
        <input
          className="register-form__field__input"
          onChange={[Function]}
          type="number"
        />
      </div>
      <div
        className="register-form__field"
      >
        <label
          className="register-form__field__label"
        >
          phone
        </label>
        <input
          className="register-form__field__input"
          onChange={[Function]}
          type="number"
        />
      </div>
    </div>
  </div>
  <div
    className="register-form__buttons-wrapper"
  >
    <Button
      className="black-button"
      icon={null}
      onClick={[Function]}
      type="submit"
    >
      create an account
    </Button>
    <Link
      to="/login"
    >
      Back to login
    </Link>
  </div>
</form>
`;
